README.md
# NEWS
**The script will soon be integrated into HiveOS**

# HiveOS NVIDIA GPU autofan v.2.4
Nvidia gpu automatic fan speed script for HiveOS Ubuntu. 

The script has an intellectual adjustment. You need to set the range of the desired temperature. During the work the script will adjust the necessary parameters. After some time after starting the fan speed will be more stable in the specified temperature range.
	
Work on mixed RIGs with AMD and NVIDIA GPU's

>If you're upgrade HiveOS via account on website or by command ```selfupgrade``` not need re-install the script.

>If you are re-install HiveOS on disk - re-install ```autofan.sh``` too(!).

[**About HiveОS**](https://goo.gl/qXNH54)

# Disclaimer
Use and change at your own risk! Not responsible for any damages or issues, changing temperature controls, fan speed, etc. might damage your computer hardwares.

# Install
1. **```cd /home/user```** (go to ```/home/user``` directory on your HiveOS).
>Check current dir by command ```pwd```. For navigation use commands : ```cd ~``` (go to home dir),  ```cd ..``` (level up), ```cd dir``` (move to dir)
2. **```wget -O autofan.sh https://git.io/vhr1n```** (download the script)
>Or download and put autofan.sh via your sftp manager

3.**```chmod +x autofan.sh```** (change permission to 755 for run)

4.**```autofan.sh```** (run the script)


Now follow the on-screen instructions step-by-step:
- setup settings (or leave default)
- run script in GHOST mode (like service)

# Settings
 - ```DELAY``` - time delay for fan control (in seconds)
 - ```MIN_SPEED``` - minimum possible fan speed (in %)
 - ```MIN_TEMP``` - GPU temperature at which the minimum coefficient is applied
 - ```MAX_TEMP``` - GPU temperature at which the maximum coefficient is applied
 - ```MIN_COEF``` - minimum coefficient
 - ```MAX_COEF``` - maximum coefficient
 - ```MINER_STOP``` - stop miner if reaching critical temperature
 - ```CRITICAL_TEMP_MINER_STOP``` - critical temperature
 - ```PL_LIMIT``` - set lower power limit if reaching critical temperature (-5 watt every time)
 - ```CRITICAL_TEMP_PL``` - critical temperature


# Run with keys
The script has the following startup keys:

**-s** 		- start setup settings (terminate by Ctrl+C). Don't need restart the script.

**-su**		- setup settings (can edit ```MIN_COEF``` and ```MAX_COEF```)

**-g** 		- running script in GHOST mode or SCREEN mode (terminate by Ctrl+C)

**-c** 		- checking current script status and current GPU temperature/fan speed, updated every DELAY time (terminate by Ctrl+C)

**-r** 		- running script in SCREEN mode (terminate by Ctrl+C)

**-k** 		- kill script

**-u**		- selfupdate

**-d**		- restore HiveOS OC settings and turn off AUTOFAN (to turn on again run script in GHOST mode)

  Run without any keys has follow steps: 
  - start setup settings
  - checking starup status
  - running script in GHOST mode or SCREEN mode
	
	**Example : ```~# autofan.sh -c```**
	
# Startup at boot
*The script create file ```xinit.user.sh``` (or check availability) and configure to run ```autofan.sh``` at startup.
Also script create ```autofan.conf``` file to store user settings.*

# Examples of temperature and fan speed (default script settings)
- 55 C : 38 %
- 60 C : 48 %
- 65 C : 65 %
- 67 C : 72 %
- 69 C : 80 %
- 72 C : 84 %
- 75 C : 97 %

# Donate

*Nicehash:* **```3JKA47P98c9JGCy3GN7qXFC2FzeuJmXuph```**

*Zec:* **```t1fP9jWyqFEni2p4i9t3byqtimsMKv1y95T```**

*ETH:* **```0xe835a7d5605a370e4750279b28f9ce0926061ea2```**

	


changelog.md
# v.2.4     2018-06-10
- restore HiveOS OC settings and turn off script (-d key)

# v.2.3.4     2018-06-09
- some speed optimization

# v.2.3.3     2018-06-06
- bug fixed

# v.2.3.2     2018-06-04
- su key added
- bug fixed

# v.2.3.1     2018-06-04
- bug fixed

# v.2.3     2018-06-04
- smart automatic fan adjustment
- reduced number of settings

# v.2.2.2   2018-05-30
- bug fix and some code optimization

# v.2.2.1   2018-05-29
- check for running one copy in GHOST mode

# v.2.2   2018-05-28
- miner stopping after reaching the critical temperature (optional)
- set lower GPU power limit after reaching the critical temperature (optional)

# v.2.1   2018-05-27
- added selfupdate (key -u)


autofan-v2.sh
#!/bin/bash

##############################################################
## Thank you very much for your support! It's very impotant!##
##														    ##
## Nicehash donate: 3JKA47P98c9JGCy3GN7qXFC2FzeuJmXuph	    ##
## Zec donate: t1fP9jWyqFEni2p4i9t3byqtimsMKv1y95T		    ##
## ETH donate: 0xe835a7d5605a370e4750279b28f9ce0926061ea2   ##
##############################################################

#VAR
DELAY=30
MIN_SPEED=20
MIN_TEMP=60
MAX_TEMP=70
MIN_COEF=80
MAX_COEF=110
#END VAR

#-----------------
VERSION="2.0"
s_name="autofan.sh"
export DISPLAY=:0

red=$(tput setf 4)
green=$(tput setf 2)
reset=$(tput sgr0)
#--------------------

#--------------USER VAR----------
function set_var {
read -p  "Enter DELAY (default 30): "
if [[ $REPLY > 0 ]]; then DELAY=$REPLY; fi; echo -n -e "${red}DELAY=$DELAY${reset}\n"
read -p  "Enter MIN_SPEED (default 20): "
if [[ $REPLY > 0 ]]; then MIN_SPEED=$REPLY; fi; echo -n -e "${red}MIN_SPEED=$MIN_SPEED${reset}\n"
read -p  "Enter MIN TEMP (default 60): "
if [[ $REPLY > 0 ]]; then MIN_TEMP=$REPLY; fi; echo -n -e "${red}MIN_TEMP=$MIN_TEMP${reset}\n"
read -p  "Enter MAX TEMP (default 70): "
if [[ $REPLY > 0 ]]; then MAX_TEMP=$REPLY; fi; echo -n -e "${red}MAX_TEMP=$MAX_TEMP${reset}\n"
read -p  "Enter MIN_COEF (default 80): "
if [[ $REPLY > 0 ]]; then MIN_COEF=$REPLY; fi; echo -n -e "${red}MIN_COEF=$MIN_COEF${reset}\n"
read -p  "Enter MAX_COEF (default 110): "
if [[ $REPLY > 0 ]]; then MAX_COEF=$REPLY; fi; echo -n -e "${red}MAX_COEF=$MAX_COEF${reset}\n"
echo "Creating config..."
if [ ! -f "/home/user/autofan.conf" ]; then
		touch /home/user/autofan.conf
fi
echo -n > /home/user/autofan.conf
echo -e "DELAY=$DELAY\nMIN_SPEED=$MIN_SPEED\nMIN_TEMP=$MIN_TEMP\nMAX_TEMP=$MAX_TEMP\nMIN_COEF=$MIN_COEF\nMAX_COEF=$MAX_COEF" >> /home/user/autofan.conf
echo "${green}[Status]: ${reset}Config created."		
}
#---------------END USER VAR--------------

#------------Check xinit.user.sh----------------
function check_run {

if [ ! -f "/home/user/xinit.user.sh" ]; then
		touch /home/user/xinit.user.sh
		chmod +x /home/user/xinit.user.sh
		echo -e "#!/bin/bash\nscreen -dmS autofan /home/user/$s_name -r" >> /home/user/xinit.user.sh
		echo -n -e "${green}[Status]: ${reset}File xinit.user.sh created.\n"
else 
		if grep -q "screen -dmS autofan /home/user/$s_name -r" /home/user/xinit.user.sh; then
					echo -n -e "${green}[Status]: ${reset}Autorun switched on.\n"
		else 
					echo -e "screen -dmS autofan /home/user/$s_name -r" >> /home/user/xinit.user.sh
					echo -n -e "${green}[Status]: ${reset}Autorun created.\n"
		fi
fi
}
#--------------END CHECK-------------

function auto_fan {

CARDS_NUM=`nvidia-smi -L | wc -l`
echo "Found ${CARDS_NUM} GPU(s)"
echo -e -n "${green}Current AUTOFAN settings:${reset}\nDELAY=$DELAY\nMIN_SPEED=$MIN_SPEED\nMIN_TEMP=$MIN_TEMP\nMAX_TEMP=$MAX_TEMP\nMIN_COEF=$MIN_COEF\nMAX_COEF=$MAX_COEF\n"
sleep 2

while true
        do
			if test -f "/home/user/autofan.conf" ; then source /home/user/autofan.conf ; fi
            echo -n -e "${green}$(date +"%d/%m/%y %T")${reset}\n"
        for ((i=0; i<$CARDS_NUM; i++))
            do
                GPU_TEMP=`nvidia-smi -i $i --query-gpu=temperature.gpu --format=csv,noheader`
                if [ $GPU_TEMP -le $MIN_TEMP ]
                    then
                        FAN_SPEED=$(($GPU_TEMP * ($MIN_COEF-($MIN_TEMP - $GPU_TEMP) * 2)/100))
						if [ $FAN_SPEED -le $MIN_SPEED ] 
						then
								FAN_SPEED=$MIN_SPEED 
						fi

                elif [[ $GPU_TEMP > $MIN_TEMP ]] && [[ $GPU_TEMP < $MAX_TEMP ]]
                    then
						FAN_SPEED=$((  $GPU_TEMP *(($GPU_TEMP - $MIN_TEMP) * 4 + $MIN_COEF)/100 ))

				elif [ $GPU_TEMP -ge $MAX_TEMP ]
                    then
						FAN_SPEED=$(( $GPU_TEMP *(($GPU_TEMP - $MAX_TEMP) * 4 + $MAX_COEF)/100 ))
                fi

		if [ $FAN_SPEED -gt 100 ]
				then
					FAN_SPEED=100
		fi
                nvidia-settings -a [gpu:$i]/GPUFanControlState=1 > /dev/null
				nvidia-settings -a [fan:$i]/GPUTargetFanSpeed=$FAN_SPEED > /dev/null
                echo "GPU${i} ${GPU_TEMP}°C -> ${FAN_SPEED}%"
       done
sleep $DELAY
done
}

function ghost_run {
read -p  "Run script in GHOST mode? (y/n) "
if [[ $REPLY = "y" ]] ;then 
			
			screen -dmS autofan /home/user/$s_name -r
			echo "Your choice is ${green}[YES]${reset}."
			echo -n -e "${green}[Status]: ${reset}Script started in GHOST mode.\n"
elif [[ $REPLY = "n" ]] ; then
							echo "Your choice is ${red}[NO]${reset}."
							read -p  "Run script in default mode? (y/n) "
							if [[ $REPLY = "y" ]] ; then $s_name -r
							else echo "Your choice is ${red}[NO]${reset}. See your later.."; exit
							fi
else 
echo "${red}[FAIL] ${reset} Please, make a choice."
ghost_run
fi

}

if test -f "/home/user/autofan.conf" ; then source /home/user/autofan.conf ; fi


if [[ $1 = "-r" ]]; then auto_fan
elif [[ $1 = "-s" ]]; then set_var
elif [[ $1 = "-g" ]]; then ghost_run
elif [[ $1 = "-c" ]]; then
		if screen -ls | grep -q "autofan"; then
		echo "${green}[Status]:${reset} The script is running."
		else echo "${green}[Status]:${reset} The script is ${red}NOT${reset} running."
		fi
		echo -e -n "${green}Current AUTOFAN settings:${reset}\nDELAY=$DELAY\nMIN_SPEED=$MIN_SPEED\nMIN_TEMP=$MIN_TEMP\nMAX_TEMP=$MAX_TEMP\nMIN_COEF=$MIN_COEF\nMAX_COEF=$MAX_COEF\n"
		CARDS_NUM=`nvidia-smi -L | wc -l`
		echo "Found ${CARDS_NUM} GPU(s):"
		echo -n -e "${green}$(date +"%d/%m/%y %T")${reset}\n"
		for ((i=0; i<$CARDS_NUM; i++))
            do
                GPU_TEMP_C=`nvidia-smi -i $i --query-gpu=temperature.gpu --format=csv,noheader`
				GPU_FAN_C=`nvidia-smi -i $i --query-gpu=fan.speed --format=csv,noheader`
				echo "GPU${i} ${GPU_TEMP_C}°C - ${GPU_FAN_C}"
		done
		
elif [[ $1 = "-k" ]]; then pkill $s_name
else 
		echo -n -e "${green}HiveOS autofan script for NVIDIA GPU.${reset} v.${VERSION}\n"
		set_var
		check_run
		ghost_run
fi

exit


autofan.v2.3.sh
#!/bin/bash

##############################################################
## Thank you very much for your support! It's very impotant!##
##														    ##
## Nicehash donate: 3JKA47P98c9JGCy3GN7qXFC2FzeuJmXuph	    ##
## Zec donate: t1fP9jWyqFEni2p4i9t3byqtimsMKv1y95T		    ##
## ETH donate: 0xe835a7d5605a370e4750279b28f9ce0926061ea2   ##
##############################################################


DELAY=30
MIN_SPEED=40
MIN_TEMP=55 
MAX_TEMP=65
MIN_COEF=80
MAX_COEF=110
MINER_STOP=1
CRITICAL_TEMP_MINER_STOP=85
PL_LIMIT=1
CRITICAL_TEMP_PL=75


VERSION="2.3"
s_name="autofan.sh"
CONF_FILE="/home/user/autofan.conf"
export DISPLAY=:0
red=$(tput setf 4)
green=$(tput setf 2)
reset=$(tput sgr0)

function set_var {
read -p  "Enter DELAY (current $DELAY): "
[ ! -z "${REPLY##*[!0-9]*}" ] && DELAY=$REPLY
echo -n -e "${red}DELAY=$DELAY${reset}\n"
read -p  "Enter MIN_SPEED (current $MIN_SPEED): "
[ ! -z "${REPLY##*[!0-9]*}" ] && MIN_SPEED=$REPLY
echo -n -e "${red}MIN_SPEED=$MIN_SPEED${reset}\n"
read -p  "Enter MAX TEMP (current $MAX_TEMP): "
[ ! -z "${REPLY##*[!0-9]*}" ] && MAX_TEMP=$REPLY
echo -n -e "${red}MAX_TEMP=$MAX_TEMP${reset}\n"
read -p  "Enter MIN TEMP (current $MIN_TEMP): "
[ ! -z "${REPLY##*[!0-9]*}" ] && MIN_TEMP=$REPLY
echo -n -e "${red}MIN_TEMP=$MIN_TEMP${reset}\n"

read -p  "Switch on MINER_STOP (1-YES/0-NO, current state $MINER_STOP): "
[ ! -z "${REPLY##*[!0-9]*}" ] && [[ $REPLY < 2 ]] && MINER_STOP=$REPLY
echo -n -e "${red}MINER_STOP=$MINER_STOP${reset}\n"
if [[ $MINER_STOP == 1 ]]; then 
		read -p  "Enter CRITICAL_TEMP_MINER_STOP (current $CRITICAL_TEMP_MINER_STOP): "
		[ ! -z "${REPLY##*[!0-9]*}" ] && CRITICAL_TEMP_MINER_STOP=$REPLY
		echo -n -e "${red}CRITICAL_TEMP_MINER_STOP=$CRITICAL_TEMP_MINER_STOP${reset}\n"
fi
read -p  "Switch on PL_LIMIT (1-YES/0-NO, current state $PL_LIMIT): "
[ ! -z "${REPLY##*[!0-9]*}" ] && [[ $REPLY < 2 ]] && PL_LIMIT=$REPLY
echo -n -e "${red}PL_LIMIT=$PL_LIMIT${reset}\n"
if [[ $PL_LIMIT == 1 ]]; then 
		read -p  "Enter CRITICAL_TEMP_PL (current $CRITICAL_TEMP_PL): "
		[ ! -z "${REPLY##*[!0-9]*}" ] && CRITICAL_TEMP_PL=$REPLY
		echo -n -e "${red}CRITICAL_TEMP_PL=$CRITICAL_TEMP_PL${reset}\n"
fi
echo "Creating config..."
[ ! -f $CONF_FILE ] &&	touch $CONF_FILE
echo -n > $CONF_FILE
echo -e "DELAY=$DELAY\nMIN_SPEED=$MIN_SPEED\nMIN_TEMP=$MIN_TEMP\nMAX_TEMP=$MAX_TEMP\nMIN_COEF=$MIN_COEF\nMAX_COEF=$MAX_COEF\nMINER_STOP=$MINER_STOP\nCRITICAL_TEMP_MINER_STOP=$CRITICAL_TEMP_MINER_STOP\nPL_LIMIT=$PL_LIMIT\nCRITICAL_TEMP_PL=$CRITICAL_TEMP_PL" >> $CONF_FILE
echo "${green}[Status]: ${reset}Config created."		
}

function check_run {

if [ ! -f "/home/user/xinit.user.sh" ]; then
		touch /home/user/xinit.user.sh
		chmod +x /home/user/xinit.user.sh
		echo -e "#!/bin/bash\nscreen -dmS autofan /home/user/$s_name -r" >> /home/user/xinit.user.sh
		echo -n -e "${green}[Status]: ${reset}File xinit.user.sh created.\n"
else 
		if grep -q "screen -dmS autofan /home/user/$s_name -r" /home/user/xinit.user.sh; then
					echo -n -e "${green}[Status]: ${reset}Autorun switched on.\n"
		else 
					echo -e "screen -dmS autofan /home/user/$s_name -r" >> /home/user/xinit.user.sh
					echo -n -e "${green}[Status]: ${reset}Autorun created.\n"
		fi
fi
}

function safe_mode {
if [[ $1 == 1 ]]; then miner start && wd start; echo "${green}[Status]: ${reset} Miner started." 
elif [[ $1 == 2 ]]; then miner stop && wd stop; echo "${red}Critical temperature! Miner STOPPED!${reset}"
fi
}

function clock_limit_mode {
gpu_info=`nvidia-smi --query-gpu=power.min_limit,power.limit --format=csv,noheader,nounits -i $i`
pl_min=`awk -F', ' '{print $1}' <<< $gpu_info`
pl_cur=`awk -F', ' '{print $2}' <<< $gpu_info`
echo -en "${green}GPU${i}${reset} Minimal possible PL: $pl_min, curent PL: $pl_cur"
new_pl=$(( `sed 's/\.[0-9]*//' <<< $pl_cur`-5 ))
if [  `sed 's/\.[0-9]*//' <<< $pl_min` -lt $new_pl  ]
	then
		echo -en "     ${red}GPU${i}-> ${reset}Try to set PL to $new_pl"
		nvidia-settings -a [gpu:0]/GPUPowerMizerMode=1 > /dev/null 2>&1 
		nvidia-smi -i $i -pl $new_pl > /dev/null 2>&1 && echo "${green}GPU${i}-> ${reset} PL was set to $new_pl"
	else echo -en "     ${red}GPU${i}: ${reset}already done or PL too low."
fi
}

function change_coef {

if [[ $1 == 0 ]]; then 
	screen_count=`screen -ls miner | grep miner | wc -l`
	[[ $screen_count > 0 ]] && [[  $MIN_COEF > 70 ]] && MIN_COEF=$(( $MIN_COEF-1 )) && MAX_COEF=$(( $MAX_COEF-1 )) || echo "Low temp"
else MIN_COEF=$(( $MIN_COEF+2 )) && MAX_COEF=$(( $MAX_COEF+2 ))
fi

#			
echo -e " Set MIN_COEF ->$MIN_COEF & MAX_COEF ->$MAX_COEF"
#
sed -i "s/\(MIN_COEF *= *\).*/\1$MIN_COEF/" $CONF_FILE && sed -i "s/\(MAX_COEF *= *\).*/\1$MAX_COEF/" $CONF_FILE
CHANGE_COEF_FLAG=

}
function auto_fan {
CARDS_NUM=`nvidia-smi -L | wc -l`
echo "Found ${CARDS_NUM} GPU(s)"
nvidia-smi -pm 1 > /dev/null 2>&1  #enable persistence mode (PM) for the driver
echo -e -n "${green}Current AUTOFAN settings:${reset}\nDELAY=$DELAY\nMIN_SPEED=$MIN_SPEED\nMIN_TEMP=$MIN_TEMP\nMAX_TEMP=$MAX_TEMP\nMIN_COEF=$MIN_COEF\nMAX_COEF=$MAX_COEF\nMINER_STOP=$MINER_STOP\nCRITICAL_TEMP_MINER_STOP=$CRITICAL_TEMP_MINER_STOP\nPL_LIMIT=$PL_LIMIT\nCRITICAL_TEMP_PL=$CRITICAL_TEMP_PL\n"
sleep 2
while true
        do
			[[ -e $CONF_FILE ]] && . $CONF_FILE
            echo -n -e "${green}$(date +"%d/%m/%y %T")${reset}\n"
        for ((i=0; i<$CARDS_NUM; i++))
            do
                GPU_TEMP=`nvidia-smi -i $i --query-gpu=temperature.gpu --format=csv,noheader`
                if [[ $GPU_TEMP -lt $MIN_TEMP ]]
                    then
						[[ CHANGE_COEF_FLAG != 1 ]] && CHANGE_COEF_FLAG=0
                        FAN_SPEED=$(($GPU_TEMP * ($MIN_COEF-($MIN_TEMP - $GPU_TEMP) * 2)/100))
						[[ $FAN_SPEED -le $MIN_SPEED ]] && FAN_SPEED=$MIN_SPEED 
						GPU_TEMP_ALL=`nvidia-smi --query-gpu=temperature.gpu --format=csv,noheader`
									GPU_TEMP_ALL=(${GPU_TEMP_ALL//\n/ })
									for gpu_count in "${GPU_TEMP_ALL[@]}"
										do
											  [[ "$gpu_count" > $MIN_TEMP ]] && MINER_STOP=0 && break 
										done
						[[ $MINER_STOP == 1 ]] && ! screen -ls | grep -q "miner" && safe_mode 1
						MINER_STOP=1
                elif [[ $GPU_TEMP -ge $MIN_TEMP  &&  $GPU_TEMP -le $MAX_TEMP ]]
                    then
					 if [[ -n $PREV_TEMP_ALL && $(( $GPU_TEMP+1 )) -eq ${PREV_TEMP_ALL[$i]} ]]; then
								FAN_SPEED=$(( ${PREV_FAN_ALL[$i]}-1 )) 
					 			# echo "     GPU${i}: FAN_SPEED->$FAN_SPEED" 
					 elif [[ $GPU_TEMP > ${PREV_TEMP_ALL[$i]} ]]; then
							FAN_SPEED=$((  $GPU_TEMP *(($GPU_TEMP - $MIN_TEMP) * 4 + $MIN_COEF)/100 ))
					 else FAN_SPEED=${PREV_FAN_ALL[i]}
					 fi

				elif [[ $GPU_TEMP -gt $MAX_TEMP ]]
                    then
						CHANGE_COEF_FLAG=1
						FAN_SPEED=$(( $GPU_TEMP *(($GPU_TEMP - $MAX_TEMP) * 4 + $MAX_COEF)/100 ))
                fi
				[[ $GPU_TEMP -ge $CRITICAL_TEMP_MINER_STOP   &&  $MINER_STOP == 1 ]] && safe_mode 2
				[[ $GPU_TEMP -ge $CRITICAL_TEMP_PL  &&  $PL_LIMIT == 1 ]] && clock_limit_mode $i
				[[ $FAN_SPEED -gt 100 ]] && FAN_SPEED=100
				nvidia-settings -a [gpu:$i]/GPUFanControlState=1 -a [fan:$i]/GPUTargetFanSpeed=$FAN_SPEED > /dev/null 2>&1 &
                echo "GPU${i} ${GPU_TEMP}°C -> ${FAN_SPEED}%"
				PREV_FAN_ALL[$i]=$FAN_SPEED
				PREV_TEMP_ALL[$i]=$GPU_TEMP
       done
	   [[ -n $CHANGE_COEF_FLAG ]] && change_coef $CHANGE_COEF_FLAG

sleep $DELAY
done
}

function ghost_run {
read -p  "Run script in GHOST mode? (y/n) "
if [[ $REPLY = "y" ]] ;then 
					session_count=`screen -ls autofan | grep autofan | wc -l`
						if [[ $session_count > 0 ]]; then
							echo -e "${red}AUTOFAN is already running${reset}"
							echo -e "Run autofan.sh -c to see the current status"
						else 
						echo "Your choice is ${green}[YES]${reset}."
						screen -dmS autofan /home/user/$s_name -r && echo -n -e "${green}[Status]: ${reset}Script started in GHOST mode.\n"
						fi
			
elif [[ $REPLY = "n" ]] ; then
							echo "Your choice is ${red}[NO]${reset}."
							read -p  "Run script in default mode? (y/n) "
							if [[ $REPLY = "y" ]] ; then $s_name -r
							else echo "Your choice is ${red}[NO]${reset}. See your later.."; exit
							fi
else 
echo "${red}[FAIL] ${reset} Please, make a choice."
ghost_run
fi
}

function selfupdate {
echo "${green}[Status]:${reset} Checking..."
new_version=`wget -q -O- https://raw.githubusercontent.com/Steambot33/HiveOS-NVIDIA-GPU-autofan-2.0/master/version | head`
if [[ $new_version != $VERSION ]] 
		then 
		echo "${green}NEW VERSION $new_version ${reset}"
		read -p  "Continue update? (y/n)"
		if [[ $REPLY = "y" ]]
				then 
				mv autofan.sh autofan.sh.old
				if wget -q https://raw.githubusercontent.com/Steambot33/HiveOS-NVIDIA-GPU-autofan-2.0/master/autofan.sh
				then 
				rm autofan.sh.old
				chmod +x autofan.sh
				echo "${green}[Status]:${reset} The script updated."
				echo "${green}[Status]:${reset} Restart script! (autofan.sh -k and autofan.sh -g)"
				else 
						echo "${red}[ FAIL ]{reset}"
						mv autofan.sh.old autofan.sh
				fi
		else echo "Your choice is ${red}[NO]${reset}."
		fi
else echo "${green}[Status]:${reset}You use actual version $new_version"
fi
}

[[ -e $CONF_FILE ]] && . $CONF_FILE 

case $1 in
	-t)
		change_coef 1
		;;
	-r)
		auto_fan
	;;
	
	-s)
		set_var
	;;
	
	-g)
		ghost_run
	;;
	
	-u)
		selfupdate
		
	;;
	
	-c)
		if screen -ls | grep -q "autofan"; then
		echo "${green}[Status]:${reset} The script is running."
		else echo "${green}[Status]:${reset} The script is ${red}NOT${reset} running."
		fi
		echo -en "${green}Current AUTOFAN settings:${reset}\nDELAY=$DELAY\nMIN_SPEED=$MIN_SPEED\nMIN_TEMP=$MIN_TEMP\nMAX_TEMP=$MAX_TEMP\nMIN_COEF=$MIN_COEF\nMAX_COEF=$MAX_COEF\nMINER_STOP=$MINER_STOP\nCRITICAL_TEMP_MINER_STOP=$CRITICAL_TEMP_MINER_STOP\nPL_LIMIT=$PL_LIMIT\nCRITICAL_TEMP_PL=$CRITICAL_TEMP_PL\n"
		CARDS_NUM=`nvidia-smi -L | wc -l`
		echo "Found ${CARDS_NUM} GPU(s):"
		while true
        do
			echo -n -e "${green}$(date +"%d/%m/%y %T")${reset}\n"
			for ((i=0; i<$CARDS_NUM; i++))
				do
					GPU_TEMP_C=`nvidia-smi -i $i --query-gpu=temperature.gpu --format=csv,noheader`
					GPU_FAN_C=`nvidia-smi -i $i --query-gpu=fan.speed --format=csv,noheader`
					echo "GPU${i} ${GPU_TEMP_C}°C - ${GPU_FAN_C}"
			done
		sleep $DELAY 
		done
	;;
	-k)
		pkill $s_name
	;;
	
	*)
	echo -n -e "${green}HiveOS autofan script for NVIDIA GPU.${reset} v.${VERSION}\n"
		set_var
		check_run
		ghost_run
	;;
esac
exit
